builtin/logical/database/version_wrapper.go

```diff
@@ -0,0 +1,272 @@
+		dbw = databaseVersionWrapper{
+			v4: legacyDB,
+		}
+		return dbw, nil
+	}
+
+	return dbw, fmt.Errorf("invalid database version")
+}
+
+// Initialize the underlying database. This is analogous to a constructor on the database plugin object.
+// Errors if the wrapper does not contain an underlying database.
+func (d databaseVersionWrapper) Initialize(ctx context.Context, req newdbplugin.InitializeRequest) (newdbplugin.InitializeResponse, error) {
+	if !d.isNewDB() && !d.isLegacyDB() {
+		return newdbplugin.InitializeResponse{}, fmt.Errorf("no underlying database specified")
+	}
+
+	// v5 Database
+	if d.isNewDB() {
+		return d.v5.Initialize(ctx, req)
+	}
+
+	// v4 Database
+	saveConfig, err := d.v4.Init(ctx, req.Config, req.VerifyConnection)
+	if err != nil {
+		return newdbplugin.InitializeResponse{}, err
+	}
+	resp := newdbplugin.InitializeResponse{
+		Config: saveConfig,
+	}
+	return resp, nil
+}
+
+// NewUser in the database. This is different from the v5 Database in that it returns a password as well.
+// This is done because the v4 Database is expected to generate a password and return it. The NewUserReponse
+// does not have a way of returning the password so this function signature needs to be different.
+// The password returned here should be considered the source of truth, not the provided password.
+// Errors if the wrapper does not contain an underlying database.
+func (d databaseVersionWrapper) NewUser(ctx context.Context, req newdbplugin.NewUserRequest) (resp newdbplugin.NewUserResponse, password string, err error) {
+	if !d.isNewDB() && !d.isLegacyDB() {
+		return newdbplugin.NewUserResponse{}, "", fmt.Errorf("no underlying database specified")
+	}
+
+	// v5 Database
+	if d.isNewDB() {
+		resp, err = d.v5.NewUser(ctx, req)
+		return resp, req.Password, err
+	}
+
+	// v4 Database
+	stmts := dbplugin.Statements{
+		Creation: req.Statements.Commands,
+		Rollback: req.RollbackStatements.Commands,
+	}
+	usernameConfig := dbplugin.UsernameConfig{
+		DisplayName: req.UsernameConfig.DisplayName,
+		RoleName:    req.UsernameConfig.RoleName,
+	}
+	username, password, err := d.v4.CreateUser(ctx, stmts, usernameConfig, req.Expiration)
+	if err != nil {
+		return resp, "", err
+	}
+
+	resp = newdbplugin.NewUserResponse{
+		Username: username,
+	}
+	return resp, password, nil
+}
+
+// UpdateUser in the underlying database. This is used to update any information currently supported
+// in the UpdateUserRequest such as password credentials or user TTL.
+// Errors if the wrapper does not contain an underlying database.
+func (d databaseVersionWrapper) UpdateUser(ctx context.Context, req newdbplugin.UpdateUserRequest, isRootUser bool) (saveConfig map[string]interface{}, err error) {
+	if !d.isNewDB() && !d.isLegacyDB() {
+		return nil, fmt.Errorf("no underlying database specified")
+	}
+
+	// v5 Database
+	if d.isNewDB() {
+		_, err := d.v5.UpdateUser(ctx, req)
+		return nil, err
+	}
+
+	// /////////////////////////////////////////////
```

Comment:
```suggestion
```

Reply:
Done

Reply:
Did you end up pushing new changes? I still see this on the latest commit.

Reply:
Just got pushed.